### PR TITLE
Change the Prefix field of the Route Information Option to be variable-length.

### PR DESCRIPTION
**Pull request by y-kzm**
This PR enables prefixes to be sent in smaller chunks within the Route Information Options.

According to RFC 4191, Section 2.3: 
_The Length field is 1, 2, or 3 depending on the Prefix Length. If Prefix Length is greater than 64, then Length must be 3. If Prefix Length is greater than 0, then Length must be 2 or 3. If Prefix Length is zero, then Length must be 1, 2, or 3._

- Before change
```
          route info option (24), length 24 (3):  2001:db8:1::1/128, pref=medium, lifetime=300s
            0x0000:  8000 0000 012c 2001 0db8 0001 0000 0000
            0x0010:  0000 0000 0001
          route info option (24), length 24 (3):  2001:db8:2::/64, pref=medium, lifetime=300s
            0x0000:  4000 0000 012c 2001 0db8 0002 0000 0000
            0x0010:  0000 0000 0000
```

- After change
```
          route info option (24), length 24 (3):  2001:db8:1::1/128, pref=medium, lifetime=300s
            0x0000:  8000 0000 012c 2001 0db8 0001 0000 0000
            0x0010:  0000 0000 0001
          route info option (24), length 16 (2):  2001:db8:2::/64, pref=medium, lifetime=300s
            0x0000:  4000 0000 012c 2001 0db8 0002 0000
```

Signed-off-by: y-kzm <kzm-yokoo@hiroshima-u.ac.jp>

### CODE REVIEW

File: send.c [hunk 1]
```diff
@@ -571,8 +571,16 @@ static struct safe_buffer_list *add_ra_options_route(struct safe_buffer_list *sb
...
+		if (route->PrefixLen == 0) {
+			rinfo.nd_opt_ri_len = 1;
+		} else if (route->PrefixLen > 0 && route->PrefixLen <= 64) {
+			rinfo.nd_opt_ri_len = 2;
+		} else if (route->PrefixLen > 64 && route->PrefixLen <= 128) {
+			rinfo.nd_opt_ri_len = 3;
+		} else {
+			route = route->next;
+			continue;
+		}
```
Comment (robbat2) [inline send.c:583]: silently dropping the route here seems weird. At the very least it should drop; but also it should be IMPOSSIBLE to reach this if we validate routes earlier.

Comment (robbat2) [reply send.c:583]: @y-kzm  bump

**Comment by Neustradamus** [comment]
@y-kzm: Have you seen @robbat2 review?

**Comment by y-kzm** [comment]
The test failed because the expected data for the prefix field of the RIO had a fixed length.
Before (expected data): `test/send.c:164`
```c
	unsigned char expected[] = {
		0x18, 0x03, 0x30, 0x18, 0x00, 0x00, 0x27, 0x10, 0xfe, 0x80, 0x00, 0x0f, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00,
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x18, 0x03, 0x28, 0x08, 0xff, 0xff, 0xff, 0xff, 0xfe, 0x80, 0x00, 0x0f,
		0x00, 0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x18, 0x03, 0x20, 0x00, 0x00, 0x00,
		0x0b, 0xb8, 0xfe, 0x80, 0x00, 0x0f, 0x00, 0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	};
```
`type`: 0x18 (24)
`length`: 0x03 (3*8=24byte)
`prefix length`: 0x30 (/48)
`Resvd|Prf|Resvd` : 0x18 (0b000 11 000)  low=0b11
`Route Lifetime`:  0x00002710 (=10000)
`Prefix` : 0xfe, 0x80, 0x00, 0x0f, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 (fe80:000f:0001:0000:0000:0000:0000:0000) // fixed
…

After (expected data): `test/send.c:164`
```c
	unsigned char expected[] = {
		0x18, 0x02, 0x30, 0x18, 0x00, 0x00, 0x27, 0x10, 0xfe, 0x80, 0x00, 0x0f, 0x00, 0x01, 0x00, 0x00,
		0x18, 0x02, 0x28, 0x08, 0xff, 0xff, 0xff, 0xff, 0xfe, 0x80, 0x00, 0x0f, 0x00, 0x02, 0x00, 0x00,
		0x18, 0x02, 0x20, 0x00, 0x00, 0x00, 0x0b, 0xb8, 0xfe, 0x80, 0x00, 0x0f, 0x00, 0x02, 0x00, 0x00,
		0x18, 0x03, 0x80, 0x00, 0x00, 0x00, 0x0b, 0xb8, 0xfe, 0x80, 0x00, 0x0f, 0x00, 0x02, 0x00, 0x00,
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	};
```
`type`: 0x18 (24)
`length`: 0x02 (2*8=16byte)
`prefix length`: 0x30 (/48)
`Resvd|Prf|Resvd` : 0x18 (0b000 11 000)  low=0b11
`Route Lifetime`:  0x00002710 (=10000)
`Prefix` : 0xfe, 0x80, 0x00, 0x0f, 0x00, 0x01, 0x00, 0x00 (fe80:000f:0001:0000:) // variable
...

With this change, I have added a test case for when the prefix length is /128 in `test/test1.conf`.
Please review and approve the workflow again.
Thank you.

**Comment by stappersg** [comment]
On Mon, Nov 25, 2024 at 02:32:41AM -0800, yykzm wrote:
> 
> With this change, I have added a test case for when the prefix length is /128 in `test/test1.conf`.
> Please review and approve the workflow again.

After approval of the workflow,
there was: All checks have passed, 40 successful checks

> Thank you.

Thank you for your contribution to radvd.


Groeten
Geert Stappers
-- 
Silence is hard to parse

**Comment by stappersg** [comment]
Oh, https://github.com/radvd-project/radvd/pull/193#issuecomment-1296286565 reminded me
on _add `Sign-off-by` trailer_

**Comment by y-kzm** [comment]
I added `Signed-off-by` trailer to the relevant commits.

**Review by robbat2 — APPROVED** [review]
